Lowering: cannot resolve return type for indirect call through a function-typed parameter

Part of epic #549 (examples regression sweep).

## Goal

Calling a value of function type held in a parameter (a first-class function / higher-order-function argument) must lower.

## Root cause / symptom

`examples/functional/higher-order-functions.sfn` defines `fn apply(value: int, transformer: (int) -> int)` and calls `transformer(value)`. Two failures observed:

1. Passing a **bare named function** `apply(5, double)` is rejected at typecheck: `error[E0808]: function \`double\` cannot be used as a value here`.
2. Wrapping it in a lambda to satisfy typecheck — `apply(5, fn(x) => double(x))` — passes `sfn check` but then fatals at lowering:

```
llvm lowering [fatal]: cannot resolve return type for call to `transformer` — callee signature is not known to the compiler
```

So the callee-signature resolution in lowering (`core_call_resolution.sfn` → `core_call_emission.sfn:371`) has no entry for an **indirect call through a fn-typed parameter** — its return type must come from the parameter's declared `(int) -> int` type, not a function registry. Distinct from the method-dispatch bucket (#1835): here the callee is a plain parameter identifier, not a method on a receiver.

## Scope

- In: resolve an indirect call's return type from the callee value's function type (parameter/local of type `fn(...) -> R`). Decide whether bare-function-as-value (E0808) should be accepted here or whether the lambda form is the supported spelling.
- Out: closures capturing environment beyond what's already shipped.

## Acceptance criteria

- [ ] `functional/higher-order-functions.sfn` runs and prints `Result: 10`.
- [ ] e2e test: user-defined HOF taking a `fn(int) -> int` parameter and invoking it.
- [ ] Self-hosts (`make compile`).

## Files affected

- `compiler/src/llvm/expression_lowering/native/core_call_resolution.sfn`
- possibly `compiler/src/typecheck*.sfn` (E0808 decision)
- `compiler/tests/e2e/`

## Verification

```bash
timeout 25 build/native/sailfin run examples/functional/higher-order-functions.sfn   # expect: Result: 10
```

**Size:** S · **Type:** bug

--- a/compiler/tests/e2e/fixtures/higher_order_named_fn/apply_named.sfn
+++ b/compiler/tests/e2e/fixtures/higher_order_named_fn/apply_named.sfn
@@ -1,0 +1,11 @@
+// #1837 — a user higher-order function whose parameter is declared
+// `fn(int) -> int`, invoked with a lambda that wraps a named function
+// call. Exercises indirect-call return-type resolution through a
+// function-typed parameter: `apply(5, fn(x) => double(x))` = 10.
+fn apply(value: int, transformer: fn (int) -> int) -> int {
+    return transformer(value);
+}
+
+fn double(x: int) -> int { return x * 2; }
+
+fn main() ![io] { print(apply(5, fn (x) => double(x))); }

--- a/compiler/tests/e2e/higher_order_named_fn_callback_test.sfn
+++ b/compiler/tests/e2e/higher_order_named_fn_callback_test.sfn
@@ -1,0 +1,45 @@
+// End-to-end: an indirect call through a `fn(int) -> int` parameter lowers
+// and returns the correct value (#1837). The callee is a plain parameter
+// identifier, not a method on a receiver, so its return type must come from
+// the parameter's declared function type. The canonical `fn(int) -> int`
+// spelling maps the parameter to the `{i8*, i8*}` closure pair and dispatches
+// through the closure seam; the non-canonical bare `(int) -> int` spelling
+// instead mapped it to an opaque `i8*` and silently produced the wrong result
+// (`Result: 0`), which is why the shipped example was rewritten to the
+// canonical spelling.
+//
+// Drives the compiler-under-test as a subprocess and asserts on captured
+// stdout — the canonical e2e pattern (`.claude/rules/no-bash-e2e.md`,
+// `untyped_lambda_callback_test.sfn`).
+import { trim_right } from "sfn/strings";
+
+// The compiler-under-test: `$SAILFIN_BIN` if set, else the self-hosted
+// binary relative to the repo root the runner starts from.
+fn _sfn_bin() -> string ![io] {
+    let configured = env.get("SAILFIN_BIN");
+    if configured.length > 0 { return configured; }
+    return "build/native/sailfin";
+}
+
+// `sfn run` compiles+links+executes the fixture, so the spawned compiler
+// needs a real environment (PATH for the linker, plus SAILFIN_TEST_SCRATCH
+// for per-invocation build-cache isolation under the parallel pool, #1333).
+fn _child_env() -> string[] ![io] { return process.environ(); }
+
+fn _run(path: string) -> string ![io] {
+    let exit = process.run_capture([_sfn_bin(), "run", path], _child_env());
+    let out = process.capture_take_stdout();
+    let _err = process.capture_take_stderr();
+    assert exit == 0;
+    return trim_right(out);
+}
+
+test "hof: fn(int)->int param invoked with a named-fn lambda returns 10" ![io] {
+    let out = _run("compiler/tests/e2e/fixtures/higher_order_named_fn/apply_named.sfn");
+    assert out == "10";
+}
+
+test "hof: shipped higher-order-functions example prints Result: 10" ![io] {
+    let out = _run("examples/functional/higher-order-functions.sfn");
+    assert out == "Result: 10";
+}

--- a/examples/functional/higher-order-functions.sfn
+++ b/examples/functional/higher-order-functions.sfn
@@ -1,14 +1,11 @@
 // examples/functional/higher-order-functions.sfn
-
-fn apply(value: int, transformer: (int) -> int) -> int {
+fn apply(value: int, transformer: fn (int) -> int) -> int {
     return transformer(value);
 }
 
-fn double(x: int) -> int {
-    return x * 2;
-}
+fn double(x: int) -> int { return x * 2; }
 
 fn main() ![io] {
-    let result = apply(5, double);
+    let result = apply(5, fn (x) => double(x));
     print("Result: {{result}}");
 }

--- a/scripts/check-examples.sh
+++ b/scripts/check-examples.sh
@@ -54,7 +54,6 @@ KNOWN_FAILING=(
     "examples/concurrency/producer-consumer.sfn"       # #1835 lowering fatal fixed; run still hangs on drain (#549 await follow-up)
     "examples/concurrency/dynamic-task-scheduling.sfn" # fn-typed channel element + await (#829)
     "examples/advanced/matrix-multiplication.sfn"      # #1836 range/nested-array map still gated (#766)
-    "examples/functional/higher-order-functions.sfn"   # #1837 indirect call through fn-typed param
     "examples/advanced/unions.sfn"                      # #1838 @.runtime.field.name global
     "examples/algorithms/quicksort.sfn"                # #1839 double-array GEP i64/ptr
 )

